File: packages/marketdata/src/tape.ts
import type { Side, TradeEvent } from '@exchange/common';

/**
 * Trade tape — bounded ring of recent trades. The UI's "tape" panel
 * shows the most recent ~50 of these.
 */

export interface TapeEntry {
  tradeId: string;
  ts: number;
  price: string;        // bigint as string
  qty: string;
  aggressor: Side;
}

export class Tape {
  private readonly ring: TapeEntry[] = [];
  constructor(private readonly capacity: number = 200) {}

  push(ev: TradeEvent): TapeEntry {
    const entry: TapeEntry = {
      tradeId: ev.tradeId,
      ts: ev.ts,
      price: ev.price.toString(),
      qty: ev.qty.toString(),
      aggressor: ev.aggressor,
    };
    this.ring.push(entry);
    if (this.ring.length > this.capacity) this.ring.shift();
    return entry;
  }

  recent(limit: number): TapeEntry[] {
    const n = Math.min(limit, this.ring.length);
    return this.ring.slice(this.ring.length - n);
  }

  clear(): void {
    this.ring.length = 0;
  }
}

File: packages-go/marketdata/tape.go
package main

import (
	"strconv"

	"github.com/KaustubhPatange/exchange/common"
)

type TapeEntry struct {
	TradeID   string      `json:"tradeId"`
	TS        int64       `json:"ts"`
	Price     string      `json:"price"`
	Qty       string      `json:"qty"`
	Aggressor common.Side `json:"aggressor"`
}

// Tape is a trade tape — bounded ring of recent trades. The UI's "tape"
// panel shows the most recent ~50 of these.
type Tape struct {
	capacity int
	ring     []TapeEntry
}

func NewTape(capacity int) *Tape {
	if capacity == 0 {
		capacity = 200
	}
	return &Tape{capacity: capacity, ring: make([]TapeEntry, 0, capacity)}
}

func (t *Tape) Push(ev *common.TradeEvent) TapeEntry {
	entry := TapeEntry{
		TradeID:   ev.TradeID,
		TS:        ev.TS,
		Price:     strconv.FormatInt(ev.Price, 10),
		Qty:       strconv.FormatInt(ev.Qty, 10),
		Aggressor: ev.Aggressor,
	}
	t.ring = append(t.ring, entry)
	if len(t.ring) > t.capacity {
		t.ring = t.ring[1:]
	}
	return entry
}

func (t *Tape) Recent(limit int) []TapeEntry {
	n := min(limit, len(t.ring))
	return t.ring[len(t.ring)-n:]
}

func (t *Tape) Clear() {
	t.ring = make([]TapeEntry, 0, t.capacity)
}
